helix-core/src/match_brackets.rs

```diff
@@ -90,3 +149,60 @@ fn surrounding_bytes(doc: &Rope, node: &Node) -> Option<(usize, usize)> {
 
     Some((start_byte, end_byte))
 }
+
+enum Range {
+    Forward(std::ops::Range<usize>),
+    Backward(std::iter::Rev<std::ops::Range<usize>>),
+}
+
+impl Iterator for Range {
+    type Item = usize;
+    fn next(&mut self) -> Option<usize> {
+        match self {
+            Range::Forward(range) => range.next(),
+            Range::Backward(range) => range.next(),
+        }
+    }
+}
```

Comment:
I think it's better to avoid introducing a custom type here. Ropeys iterators are extremly flexible and allow both a more performant and more correct implementation.
Simply use `doc.chars_at(cursor_pos).reversed()`  and `doc.chars_at(cursor_pos + 1)`  for reverse/forward iteration respectively. These both return the same type so no need for an extra enum. You can either use `.enumerate()` to track your position in the iterator and return `char_pos - i - 1` in the reverse case and `char_pos + 1 + i` in the forward case

Reply:
Yeah, I did not realise how simple it is using the rope `chars_at`.